pkg/operator/rotate.go

```diff
@@ -0,0 +1,242 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to create intermediate certificate signed by previous ca: %v", err)
+	}
+
+	return &signingCA{
+		config:             newCAConfig,
+		bundle:             bundle,
+		intermediateCACert: newCACertSignedByOldCA,
+		oldCAExpiry:        currentCACertSignedByNewCA.NotAfter,
+	}, nil
+}
+
+// createIntermediateCACert creates a new intermediate CA cert from a template provided by
+// the target CA cert and issued by the signing cert. This ensures that certificates
+// issued by the target CA can be trusted by clients that trust the signing CA.
+func createIntermediateCACert(targetCACert, signingCACert *x509.Certificate, signingKey *rsa.PrivateKey, expiry *time.Time) (*x509.Certificate, error) {
+	// Copy the target cert to allow modification.
+	template, err := x509.ParseCertificate(targetCACert.Raw)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to copy ca certificate: %v", err)
+	}
+	// Enable key identity chaining
+	template.AuthorityKeyId = signingCACert.SubjectKeyId
+
+	// Update the expiry if necessary
+	if expiry != nil {
+		template.NotAfter = *expiry
+	}
+
+	caCert, err := util.CreateCertificate(template, signingCACert, targetCACert.PublicKey, signingKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating intermediate CA certificate: %v", err)
+	}
+
+	return caCert, nil
+}
+
+// certHalfwayExpired indicates whether half of the cert validity period has elapsed.
+func certHalfwayExpired(cert *x509.Certificate) bool {
+	halfValidPeriod := cert.NotAfter.Sub(cert.NotBefore).Nanoseconds() / 2
+	halfExpiration := cert.NotBefore.Add(time.Duration(halfValidPeriod) * time.Nanosecond)
+	return time.Now().After(halfExpiration)
+}
+
+// forceRotationReason attempts to retrieve a force rotation reason
+// from the provided raw UnsupportedConfigOverrides.
+func forceRotationReason(rawUnsupportedServiceCAConfig []byte) (string, error) {
+	serviceCAConfig := &unsupportedServiceCAConfig{}
+	if raw := rawUnsupportedServiceCAConfig; len(raw) > 0 {
+		jsonRaw, err := kyaml.ToJSON(raw)
+		if err != nil {
+			klog.Warning(err)
+			// maybe it's just json
+			jsonRaw = raw
+		}
+		if err := json.Unmarshal(jsonRaw, serviceCAConfig); err != nil {
+			return "", err
+		}
+	}
+
+	return serviceCAConfig.ForceRotation.Reason, nil
+}
+
+// RawUnsupportedServiceCAConfig returns the raw value of the operator field
+// UnsupportedConfigOverrides for the given force rotation reason.
+func RawUnsupportedServiceCAConfig(reason string) ([]byte, error) {
+	config := &unsupportedServiceCAConfig{
+		ForceRotation: forceRotationConfig{
+			Reason: reason,
+		},
+	}
+	return json.Marshal(config)
+}
+
+// forcedRotationRequired indicates whether the force rotation reason is not empty and
+// does not match the annotation stored on the signing secret.
+func forcedRotationRequired(secret *corev1.Secret, reason string) bool {
+	if len(reason) == 0 {
+		return false
+	}
+	if secret.Annotations == nil {
+		return true
+	}
```

Comment:
nit: you can skip this if

Reply:
Will `Annotations` will never be nil on a resource returned from the API?  In unit testing that won't be true unless the map is explicitly initialized, but if that's testing-only I should update initialization of secrets used for testing.

Reply:
`nil["foo"]` gives empty string. Doesn't that match your logic here?

Reply:
I was confused - I added this check some time ago in response to an error I saw but it was clearly unrelated.

Done.